### DISCUSSION TITLE
Wave throws `401 Unauthorized` error when deployed on Heroku

### DISCUSSION THREAD

**Discussion by Mathanraj-Sharma** [h2oai/wave#1524] [Q&A] [answered]
Hi Team, I tried deploying a wave app to Heroku using the git method as explained [here](https://devcenter.heroku.com/articles/getting-started-with-python#deploy-the-app). 

I used a `Procfile` to start the app which contains below web target

```
web: H2O_WAVE_ADDRESS="http://127.0.0.1:${PORT}" H2O_WAVE_LISTEN=":${PORT}" wave run --no-reload ai_app/app.py
```

App wave server & app started as expected. But in a few seconds wave app throws a `401 Unauthorized` error.  Any help on debugging it will be appreciated 

![Screen Shot 2022-06-29 at 4 56 20 AM](https://user-images.githubusercontent.com/30664910/176321779-be5b16be-cb89-4b8b-ba2c-3256e86db4b7.png)

**Comment by mturoci** [comment]
The logs say you started 2 apps, one works fine, the other gets the auth error. Do you intend to start 2 of them? If not, I would start there.

**Reply by Mathanraj-Sharma** [reply]
@mturoci I am not intended to start two apps, I run the below command

`H2O_WAVE_ADDRESS="http://127.0.0.1:${PORT}" H2O_WAVE_LISTEN=":${PORT}" wave run --no-reload ai_app/app.py`

but it starts two apps, I don't know why. Is there any way to tell the wave to start only a single instance of the app?

**Reply by mturoci** [reply]
wave run always starts a single app only. The problem is on Heroku's side, as pointed out in the comment below.

**Comment by Mathanraj-Sharma** [comment, marked as answer]
@mturoci Thanks for pointing out the first step to debug it,

I found out what is wrong, when the web process is called it starts two instances of the app. I guess, by default, Heroku has a configuration to start two workers.

I set the `WEB_CONCURRENCY=1` and it works fine now.

we just have to add this in Procfile

`web: WEB_CONCURRENCY=1 H2O_WAVE_ADDRESS="http://127.0.0.1/:${PORT}" H2O_WAVE_LISTEN=":${PORT}" wave run --no-reload ai_app/app.py`

**Reply by Mathanraj-Sharma** [reply]
more on this article
https://mathanrajsharma.medium.com/how-to-deploy-h2o-wave-app-on-heroku-2abe2240731c
